js/materialbox.js

```diff
@@ -16,14 +16,42 @@
       var outDuration = 200;
       var origin = $(this);
       var placeholder = $('<div></div>').addClass('material-placeholder');
+      placeholder.css("display","table"); // for the progress bar to be same width of img
+      var progressWrapper = $("<div style='position: relative;bottom: 12px;'></div>");
+      var progress = $('<div></div>').addClass('progress');
+      var indeterminateProgress = $('<div></div>').addClass('indeterminate');
+      progress.append(indeterminateProgress);
+      progressWrapper.append(progress);
+
       var originalWidth = 0;
       var originalHeight = 0;
       var ancestorsChanged;
       var ancestor;
       origin.wrap(placeholder);
 
+      var imageChanged = false;
+      // thumbnail data for switching back from original in returnToOriginal
+      var originalSrc = origin.attr("src");
+      var originalSrcWidth = origin.css("width");
+      var originalSrcHeight = origin.css("height");
+
+
 
       origin.on('click', function(){
+        if (!imageChanged && origin.data('external')){
+          // if thumbnail has original version of image
+          progressWrapper.insertAfter(origin); // show progress while new src loads
+          origin.load(function(){ // will enlarge image once new src is loaded
+            enlargeImage(); // enlarge loaded src as normal
```

Comment:
There should be also added:

```
            originalWidth = origin.width();
            originalHeight = origin.height();
            origin.css({
              width: originalSrcWidth,
              height: originalSrcHeight
            });
```
So the loaded image will be resized right after loading to the thumbnail size and the resized to full size with overlay. Makes animation look little bit more smooth. Still placeholder changes it's size for a moment. I still think about a workaround.